src/unity/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -507,18 +529,87 @@ void activity_classifier::perform_training_iteration() {
   float average_batch_accuracy = cumulative_batch_accuracy / num_batches;
 
   // Report progress if we have an active table printer.
-  // TODO: Report validation metrics.
-  
-  if (training_table_printer_) {
-    training_table_printer_->print_progress_row(
-        iteration_idx, iteration_idx + 1, average_batch_accuracy, average_batch_loss, progress_time());
+  if (validation_data_iterator_ != nullptr) {
+
+    float cumulative_val_accuracy = 0.f;
+    float cumulative_val_loss = 0.f;
+    size_t val_size = 0;
+
+    validation_data_iterator_->reset();
+
+    while (validation_data_iterator_->has_next_batch()) {
+      data_iterator::batch val =
+          validation_data_iterator_->next_batch(batch_size);
+      std::map<std::string, shared_float_array> results =
+          training_model_->predict({
+              {"input", val.features},
+              {"labels", val.labels},
+              {"weights", val.weights},
+          });
+
+      const shared_float_array &val_output = results.at("output");
+
+      shared_float_array loss = results.at("loss");
```

Comment:
Nit: `val_output` and `loss` should logically have the same type, either both a const reference or both just a `shared_float_array` value. May as well have them both be references.